FEXCore/Source/Interface/IR/Passes/ConstProp.cpp

```diff
@@ -542,7 +543,8 @@ bool ConstProp::ConstantPropagation(IREmitter *IREmit, const IRListView& Current
       auto Op = IROp->CW<IR::IROp_StoreMem>();
       auto AddressHeader = IREmit->GetOpHeader(Op->Addr);
 
-      if (AddressHeader->Op == OP_ADD && AddressHeader->Size == 8) {
+      if (AddressHeader->Op == OP_ADD &&
+          (AddressHeader->Size == 8 || AddressHeader->Size == 4)) {
```

Comment:
Additional note here is that the Size check here should be checking operating mode size instead of 4 or 8.
This is because x86 is weird and allows you to change the address size of arbitrary operations. 16-bit, 32-bit, and 64-bit are all allowed.
When running in 64-bit mode, only allow this optimization for 8-byte sized addresses, when operating in 32-bit mode only allow 4-byte sized addresses.

An example where this would be an issue is a 64-bit application using 32-bit addressing to zext the lower 32-bits with garbage in the upper bits.

16-bit addressing can basically be ignored since Linux reserves the lowest 16KB so nothing uses it (we also don't support 16-bit x86).

Reply:
Thanks for commenting further. I am still working on improving this but mostly focusing on the details of what arm and intel allows as addressing modes since they seem to be two very different things. I hope to have a further work here ready this week.